File: libs/moqtail-ts/src/model/control/publish_namespace_done.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { Tuple } from '../common/tuple'
import { ControlMessageType } from './constant'
import { LengthExceedsMaxError } from '../error/error'

/**
 * @public
 * Represents a protocol PublishNamespaceDone message, used to unpublish a track namespace.
 */
export class PublishNamespaceDone {
  /**
   * Constructs a PublishNamespaceDone message.
   * @param trackNamespace - The track namespace to unpublish.
   */
  constructor(public readonly trackNamespace: Tuple) {}

  /**
   * Gets the control message type for this PublishNamespaceDone message.
   * @returns The ControlMessageType.PublishNamespaceDone enum value.
   */
  getType(): ControlMessageType {
    return ControlMessageType.PublishNamespaceDone
  }

  /**
   * Serializes the PublishNamespaceDone message to a frozen byte buffer.
   * @returns The serialized message as a FrozenByteBuffer.
   * @throws :{@link LengthExceedsMaxError} If the payload exceeds 65535 bytes.
   */
  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.PublishNamespaceDone)
    const payload = new ByteBuffer()
    payload.putTuple(this.trackNamespace)
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError(
        'PublishNamespaceDone::serialize(payloadBytes.length)',
        0xffff,
        payloadBytes.length,
      )
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  /**
   * Parses a PublishNamespaceDone message payload from a buffer.
   * @param buf - The buffer containing the payload.
   * @returns The parsed PublishNamespaceDone message.
   */
  static parsePayload(buf: BaseByteBuffer): PublishNamespaceDone {
    const trackNamespace = buf.getTuple()
    return new PublishNamespaceDone(trackNamespace)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('PublishNamespaceDone', () => {
    test('roundtrip', () => {
      const trackNamespace = Tuple.fromUtf8Path('un/announce/me')
      const msg = new PublishNamespaceDone(trackNamespace)
      const frozen = msg.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.PublishNamespaceDone))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = PublishNamespaceDone.parsePayload(frozen)
      expect(deserialized.trackNamespace.equals(msg.trackNamespace)).toBe(true)
      expect(frozen.remaining).toBe(0)
    })
    test('excess roundtrip', () => {
      const trackNamespace = Tuple.fromUtf8Path('un/announce/me')
      const msg = new PublishNamespaceDone(trackNamespace)
      const serialized = msg.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.PublishNamespaceDone))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = PublishNamespaceDone.parsePayload(frozen)
      expect(deserialized.trackNamespace.equals(msg.trackNamespace)).toBe(true)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })
    test('partial message', () => {
      const trackNamespace = Tuple.fromUtf8Path('un/announce/me')
      const msg = new PublishNamespaceDone(trackNamespace)
      const serialized = msg.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const buf = new ByteBuffer()
      buf.putBytes(partial)
      const frozen = buf.freeze()
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        PublishNamespaceDone.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/publish_namespace_done.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use bytes::{BufMut, Bytes, BytesMut};

use super::control_message::ControlMessageTrait;
use crate::model::common::tuple::Tuple;
use crate::model::common::varint::BufMutVarIntExt;
use crate::model::control::constant::ControlMessageType;
use crate::model::error::ParseError;

#[derive(Debug, Clone, PartialEq)]
pub struct PublishNamespaceDone {
  pub track_namespace: Tuple,
}
impl PublishNamespaceDone {
  pub fn new(track_namespace: Tuple) -> Self {
    PublishNamespaceDone { track_namespace }
  }
}
impl ControlMessageTrait for PublishNamespaceDone {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::PublishNamespaceDone)?;

    let mut payload = BytesMut::new();
    payload.extend_from_slice(&self.track_namespace.serialize()?);

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "PublishNamespaceDone::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;
    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let track_namespace = Tuple::deserialize(payload)?;
    Ok(Box::new(PublishNamespaceDone { track_namespace }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::PublishNamespaceDone
  }
}

#[cfg(test)]
mod tests {
  use crate::model::common::varint::BufVarIntExt;

  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let track_namespace = Tuple::from_utf8_path("un/announce/me");
    let publish_namespace_done = PublishNamespaceDone { track_namespace };
    let mut buf = publish_namespace_done.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishNamespaceDone as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = PublishNamespaceDone::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, publish_namespace_done);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let track_namespace = Tuple::from_utf8_path("un/announce/me");
    let publish_namespace_done = PublishNamespaceDone { track_namespace };

    let serialized = publish_namespace_done.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishNamespaceDone as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = PublishNamespaceDone::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, publish_namespace_done);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let track_namespace = Tuple::from_utf8_path("un/announce/me");
    let publish_namespace_done = PublishNamespaceDone { track_namespace };
    let mut buf = publish_namespace_done.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishNamespaceDone as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = PublishNamespaceDone::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
